core/emulate.c

```diff
@@ -0,0 +1,908 @@
+{
+    op->type = OP_NONE;
+}
+
+static void decode_op_modrm_reg(em_context_t *ctxt,
+                                em_operand_t *op)
+{
+    op->type = OP_REG;
+    op->size = ctxt->operand_size;
+    op->reg.index = ctxt->modrm.reg | (ctxt->rex.r << 3);
+}
+
+static void decode_op_modrm_rm(em_context_t *ctxt,
+                               em_operand_t *op)
+{
+    // Register operand
+    if (ctxt->modrm.mod == 3) {
+        op->type = OP_REG;
+        op->size = ctxt->operand_size;
+        op->reg.index = ctxt->modrm.rm | (ctxt->rex.b << 3);
+        return;
+    }
+
+    // Memory operand
+    op->type = OP_MEM;
+    op->size = ctxt->operand_size;
+    op->mem.ea = 0;
+    op->mem.seg = SEG_DS;
+
+    if (ctxt->address_size == 2) {
+        /* Intel SDM Vol. 2A:
+         * Table 2-1. 16-Bit Addressing Forms with the ModR/M Byte */
+        switch (ctxt->modrm.rm) {
+        case 0:
+            op->mem.ea = BX + SI;
+            break;
+        case 1:
+            op->mem.ea = BX + DI;
+            break;
+        case 2:
+            op->mem.ea = BP + SI;
+            op->mem.seg = SEG_SS;
+            break;
+        case 3:
+            op->mem.ea = BP + DI;
+            op->mem.seg = SEG_SS;
+            break;
+        case 4:
+            op->mem.ea = SI;
+            break;
+        case 5:
+            op->mem.ea = DI;
+            break;
+        case 6:
+            if (ctxt->modrm.mod == 0) {
+                op->mem.ea = insn_fetch_u16(ctxt);
+            } else {
+                op->mem.ea = BP;
+                op->mem.seg = SEG_SS;
+            }
+            break;
+        case 7:
+            op->mem.ea = BX;
+            break;
+        }
+
+        // Displacement
+        if (ctxt->modrm.mod == 1) {
+            op->mem.ea += insn_fetch_s8(ctxt);
+        } else if (ctxt->modrm.mod == 2) {
+            op->mem.ea += insn_fetch_u16(ctxt);
+        }
+    } else if (ctxt->address_size == 4 || ctxt->address_size == 8) {
+        /* Intel SDM Vol. 2A:
+         * Table 2-2. 32-Bit Addressing Forms with the ModR/M Byte */
+        if (ctxt->modrm.rm == 4) {
+            uint32_t reg_base;
+            uint32_t reg_index;
+
+            /* Intel SDM Vol. 2A:
+             * Table 2-3. 32-Bit Addressing Forms with the SIB Byte */
+            ctxt->sib.value = insn_fetch_u8(ctxt);
+            reg_base  = ctxt->sib.base  | (ctxt->rex.b << 3);
+            reg_index = ctxt->sib.index | (ctxt->rex.x << 3);
+            if ((reg_base & 7) == 5 && ctxt->modrm.mod == 0) {
+                op->mem.ea += insn_fetch_s32(ctxt);
+            } else {
+                op->mem.ea += READ_GPR(reg_base, ctxt->address_size);
+            }
+            if (reg_index != 4) {
+                uint8_t scale = 1 << ctxt->sib.scale;
+                op->mem.ea += READ_GPR(reg_index, ctxt->address_size) * scale;
+            }
+        } else if (ctxt->modrm.mod == 0 && ctxt->modrm.rm == 5) {
+            op->mem.ea += insn_fetch_s32(ctxt);
```

Comment:
After giving this more thought, I now believe we should use `_s32` when `address_size == 8`, but `_u32` when `address_size == 4`, because `ea` is a segment offset which must be non-negative (0 <= `ea` <= segment limit). Of course, we could also keep `_s32` and truncate `ea` to `address_size`.